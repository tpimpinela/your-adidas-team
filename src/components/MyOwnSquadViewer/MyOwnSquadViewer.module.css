.squad {
  margin: 2rem 0;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
  gap: 24px;

  & > div {
    max-width: 250px;
  }
}

.actions {
  display: flex;
  flex-direction: row-reverse;
}
