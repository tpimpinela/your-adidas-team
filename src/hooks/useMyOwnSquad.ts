import { useCallback, useContext } from "react";
import { MyOwnSquadContext } from "../contexts/MyOwnSquad.context";

const useMyOwnSquad = () => {
  const myOwnSquad = useContext(MyOwnSquadContext);

  const isMemberOnMySquad = useCallback(
    (teamMemberId: number) => {
      return Boolean(myOwnSquad.find(({ id }) => id === teamMemberId));
    },
    [myOwnSquad]
  );

  return {
    squad: myOwnSquad,
    isMemberOnMySquad,
  };
};

export default useMyOwnSquad;
