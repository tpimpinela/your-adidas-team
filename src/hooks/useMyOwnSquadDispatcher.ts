import { useCallback, useContext } from "react";
import { MyOwnSquadDispatchContext } from "../contexts/MyOwnSquad.context";
import { TeamMember } from "../models/teamMember.model";
import { MyOwnSquadActionType } from "../reducers/myOwnSquad.reducer";

const useMyOwnSquadDispatcher = () => {
  const dispatch = useContext(MyOwnSquadDispatchContext);

  const addTeamMember = useCallback(
    (teamMember: TeamMember) => {
      dispatch({
        type: MyOwnSquadActionType.Add,
        payload: teamMember,
      });
    },
    [dispatch]
  );

  const deleteTeamMember = useCallback(
    (id: number) => {
      dispatch({
        type: MyOwnSquadActionType.Delete,
        payload: id,
      });
    },
    [dispatch]
  );

  return {
    addTeamMember,
    deleteTeamMember,
  };
};

export default useMyOwnSquadDispatcher;
