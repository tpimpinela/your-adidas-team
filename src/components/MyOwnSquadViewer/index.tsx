import { useCallback } from "react";
import { useNavigate } from "react-router-dom";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";
import useMyOwnSquadDispatcher from "../../hooks/useMyOwnSquadDispatcher";
import { TeamMember as TeamMemberModel } from "../../models/teamMember.model";
import Button from "../Button";
import TeamMember from "../TeamMember";
import styles from "./MyOwnSquadViewer.module.css";

const MyOwnSquadViewer = () => {
  const { squad } = useMyOwnSquad();
  const { deleteTeamMember } = useMyOwnSquadDispatcher();
  const navigate = useNavigate();

  const handleTeamMemberClick = useCallback(
    (teamMember: TeamMemberModel) => {
      deleteTeamMember(teamMember.id);
    },
    [deleteTeamMember]
  );

  const handleSelectPlayersClick = useCallback(() => {
    navigate("/player-selection");
  }, [navigate]);

  return (
    <>
      <div className={styles.squad}>
        {squad.map((teamMember) => (
          <TeamMember
            teamMember={teamMember}
            onTeamMemberClicked={handleTeamMemberClick}
            key={teamMember.id}
            hideIncludedMark
            showDeleteMark
          />
        ))}
      </div>
      <div className={styles.actions}>
        <Button onClick={handleSelectPlayersClick}>Select players</Button>
      </div>
    </>
  );
};

export default MyOwnSquadViewer;
