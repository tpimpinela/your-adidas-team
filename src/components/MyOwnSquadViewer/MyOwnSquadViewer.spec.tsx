import { render, RenderResult } from "@testing-library/react";
import { beforeEach, describe, expect, it, Mock, vi } from "vitest";
import MyOwnSquadViewer from ".";
import userEvent from "@testing-library/user-event";
import { useNavigate } from "react-router-dom";
import useMyOwnSquad from "../../hooks/useMyOwnSquad";
import squadMock from "../../mocks/squads.mock";
import useMyOwnSquadDispatcher from "../../hooks/useMyOwnSquadDispatcher";

vi.mock("react-router-dom", () => ({
  useNavigate: vi.fn(),
}));
vi.mock("../../hooks/useMyOwnSquad", () => ({
  default: vi.fn(),
}));
vi.mock("../../hooks/useMyOwnSquadDispatcher", () => ({
  default: vi.fn(),
}));
vi.mock("../TeamMember", () => ({
  default: ({
    onTeamMemberClicked,
  }: {
    onTeamMemberClicked: (teamMember: any) => void;
  }) => (
    <button
      data-testid="team-member"
      onClick={() => onTeamMemberClicked({} as any)}
    >
      Click me
    </button>
  ),
}));

useMyOwnSquad;
describe("<MyOwnSquadViewer />", () => {
  let renderResult: RenderResult;
  let useNavigateMock: Mock<any, any>;
  const myOwnSquadDispatcherMock = useMyOwnSquadDispatcher as Mock<any, any>;
  const useMyOwnSquadMock = useMyOwnSquad as Mock<any, any>;
  let deleteTeamMemberMock: Mock<any, any>;

  beforeEach(() => {
    deleteTeamMemberMock = vi.fn();
    useMyOwnSquadMock.mockReturnValue({
      squad: [squadMock.players[0] as any],
    });
    myOwnSquadDispatcherMock.mockReturnValue({
      deleteTeamMember: deleteTeamMemberMock,
    });
    useNavigateMock = vi.fn();
    (useNavigate as Mock<any, any>).mockReturnValue(useNavigateMock);
    renderResult = render(<MyOwnSquadViewer />);
  });

  it("should match the snapshot", () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it("should navigate to players selection on button click", async () => {
    await userEvent.click(renderResult.getByTestId("button-component"));
    expect(useNavigateMock).toHaveBeenCalledWith("/player-selection");
  });

  it("should delete player if player is clicked", async () => {
    await userEvent.click(renderResult.getByTestId("team-member"));
    expect(deleteTeamMemberMock).toHaveBeenCalled();
  });
});
