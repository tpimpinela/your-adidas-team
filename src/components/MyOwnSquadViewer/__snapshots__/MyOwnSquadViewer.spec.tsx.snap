// Vitest Snapshot v1

exports[`<MyOwnSquadViewer /> > should match the snapshot 1`] = `
<div>
  <div
    class="_loader-container_bf668d"
  >
    <div
      class="_loader_bf668d"
    />
    <span>
      Loading...
    </span>
  </div>
  <div
    class="_squad_e50473"
  >
    <button
      data-testid="team-member"
    >
      Click me
    </button>
  </div>
  <div
    class="_actions_e50473"
  >
    <button
      class="_button_2a1efe"
      data-testid="button-component"
    >
      Select players
    </button>
  </div>
</div>
`;
